File: web/src/features/editor/domain/document-v2.ts
import type {
  EditorCrop,
  EditorDocumentV1,
  EditorObject,
  EditorTransform,
} from './document'

export type EditorEffectV2 = {
  type: 'exposure' | 'contrast' | 'saturation' | 'temperature'
  version: 1
  enabled: boolean
  parameters: Record<string, number>
}

export type EditorShapeMaskV2 = {
  type: 'rectangle' | 'ellipse'
  x: number
  y: number
  width: number
  height: number
  inverted: boolean
}

export type EditorPixelMaskV2 = {
  resource_id: string
  version: number
}

export type EditorNodeV2 = {
  id: string
  type: 'raster' | 'group' | 'adjustment'
  name?: string
  parent_id: string | null
  order_key: string
  transform: EditorTransform
  opacity: number
  blend_mode:
    'normal' | 'multiply' | 'screen' | 'overlay' | 'darken' | 'lighten'
  visible: boolean
  locked: boolean
  mask_id?: string
  asset_id?: string
  crop?: EditorCrop
  effects?: EditorEffectV2[]
  target_id?: string
  shape_mask?: EditorShapeMaskV2
  pixel_mask?: EditorPixelMaskV2
}

export type EditorDocumentV2 = {
  schema_version: 2
  renderer_semantics_version: 1
  canvas: { width: number; height: number }
  nodes: EditorNodeV2[]
}

export class UnsupportedEditorSemanticsError extends Error {
  constructor() {
    super('The document uses V2 semantics that the V1 renderer cannot preserve')
    this.name = 'UnsupportedEditorSemanticsError'
  }
}

export function migrateEditorDocumentV1ToV2(
  document: EditorDocumentV1,
): EditorDocumentV2 {
  const objects = [...document.objects].sort((a, b) => a.z_index - b.z_index)
  return {
    schema_version: 2,
    renderer_semantics_version: 1,
    canvas: { ...document.canvas },
    nodes: objects.map((object, index) => ({
      id: object.id,
      type: 'raster',
      name: object.name,
      parent_id: null,
      order_key: index.toString().padStart(8, '0'),
      transform: [...object.transform],
      opacity: object.opacity,
      blend_mode: 'normal',
      visible: object.visible,
      locked: object.locked,
      asset_id: object.asset_id,
      crop: object.crop ? { ...object.crop } : undefined,
      effects: [],
    })),
  }
}

export function compileEditorDocumentV2ToV1(
  document: EditorDocumentV2,
): EditorDocumentV1 {
  if (validateEditorDocumentV2(document).length > 0) {
    throw new TypeError('Invalid editor document V2')
  }
  if (
    document.nodes.some(
      (node) =>
        node.type !== 'raster' ||
        node.parent_id !== null ||
        node.mask_id !== undefined ||
        node.target_id !== undefined ||
        node.shape_mask !== undefined ||
        node.pixel_mask !== undefined ||
        node.blend_mode !== 'normal' ||
        (node.effects?.length ?? 0) > 0 ||
        !node.asset_id,
    )
  ) {
    throw new UnsupportedEditorSemanticsError()
  }
  const nodes = [...document.nodes].sort(
    (a, b) =>
      a.order_key.localeCompare(b.order_key) || a.id.localeCompare(b.id),
  )
  return {
    schema_version: 1,
    canvas: { ...document.canvas },
    objects: nodes.map<EditorObject>((node, index) => ({
      id: node.id,
      name: node.name,
      asset_id: node.asset_id!,
      transform: [...node.transform],
      opacity: node.opacity,
      visible: node.visible,
      locked: node.locked,
      z_index: index,
      crop: node.crop ? { ...node.crop } : undefined,
    })),
  }
}

export function validateEditorDocumentV2(document: EditorDocumentV2): string[] {
  const errors: string[] = []
  const pixels = document.canvas.width * document.canvas.height
  if (
    !Number.isInteger(document.canvas.width) ||
    !Number.isInteger(document.canvas.height) ||
    document.canvas.width < 1 ||
    document.canvas.height < 1 ||
    document.canvas.width > 8192 ||
    document.canvas.height > 8192 ||
    pixels > 36_000_000
  )
    errors.push('canvas')
  if (document.nodes.length > 500) errors.push('nodes')

  const nodes = new Map(document.nodes.map((node) => [node.id, node]))
  if (nodes.size !== document.nodes.length) errors.push('duplicate-id')
  const orderKeys = new Set<string>()
  for (const node of document.nodes) {
    if (
      !node.id ||
      node.id.length > 64 ||
      node.order_key.length < 1 ||
      node.order_key.length > 64 ||
      !/^[A-Za-z0-9]+$/.test(node.order_key)
    )
      errors.push(`identity:${node.id}`)
    const orderIdentity = `${node.parent_id === null ? '\u0000' : `\u0001${node.parent_id}`}\u0000${node.order_key}`
    if (orderKeys.has(orderIdentity)) errors.push(`order:${node.id}`)
    orderKeys.add(orderIdentity)
    if (
      !validTransform(node.transform) ||
      !Number.isFinite(node.opacity) ||
      node.opacity < 0 ||
      node.opacity > 1 ||
      ![
        'normal',
        'multiply',
        'screen',
        'overlay',
        'darken',
        'lighten',
      ].includes(node.blend_mode)
    )
      errors.push(`geometry:${node.id}`)
    if (node.type === 'raster') {
      if (
        !node.asset_id ||
        node.target_id !== undefined ||
        !validCrop(node.crop) ||
        !validShapeMask(node.shape_mask) ||
        !validPixelMask(node.pixel_mask) ||
        (node.shape_mask !== undefined &&
          (node.crop !== undefined ||
            node.mask_id !== undefined ||
            node.pixel_mask !== undefined)) ||
        (node.pixel_mask !== undefined && node.mask_id !== undefined) ||
        !validEffects(node.effects ?? [])
      )
        errors.push(`asset:${node.id}`)
    } else if (node.type === 'group') {
      if (
        node.asset_id !== undefined ||
        node.crop !== undefined ||
        node.target_id !== undefined ||
        node.shape_mask !== undefined ||
        node.pixel_mask !== undefined ||
        (node.effects?.length ?? 0) > 0
      )
        errors.push(`group:${node.id}`)
    } else if (
      node.asset_id !== undefined ||
      node.crop !== undefined ||
      node.mask_id !== undefined ||
      node.shape_mask !== undefined ||
      node.pixel_mask !== undefined ||
      node.target_id === undefined ||
      node.target_id === node.id ||
      node.blend_mode !== 'normal' ||
      node.transform.some(
        (value, index) => value !== [1, 0, 0, 1, 0, 0][index],
      ) ||
      !validEffects(node.effects ?? [])
    )
      errors.push(`adjustment:${node.id}`)
    if (node.parent_id !== null && nodes.get(node.parent_id)?.type !== 'group')
      errors.push(`parent:${node.id}`)
    if (
      node.mask_id !== undefined &&
      nodes.get(node.mask_id)?.type !== 'raster'
    )
      errors.push(`mask:${node.id}`)
    if (node.type === 'adjustment') {
      const target = nodes.get(node.target_id!)
      if (
        target?.type !== 'raster' ||
        target.parent_id !== node.parent_id ||
        document.nodes.some((candidate) => candidate.mask_id === target.id)
      )
        errors.push(`target:${node.id}`)
    }
    if (
      node.shape_mask !== undefined &&
      document.nodes.some((candidate) => candidate.mask_id === node.id)
    )
      errors.push(`shape-mask-source:${node.id}`)
    if (
      node.pixel_mask !== undefined &&
      document.nodes.some((candidate) => candidate.mask_id === node.id)
    )
      errors.push(`pixel-mask-source:${node.id}`)
    if (
      !validAncestry(node, nodes, 'parent_id', 32) ||
      !validAncestry(node, nodes, 'mask_id', 500)
    )
      errors.push(`cycle:${node.id}`)
  }
  return errors
}

function validCrop(crop?: EditorCrop) {
  if (!crop) return true
  return (
    [crop.x, crop.y, crop.width, crop.height].every(Number.isFinite) &&
    crop.x >= 0 &&
    crop.y >= 0 &&
    crop.width > 0 &&
    crop.height > 0 &&
    crop.x + crop.width <= 1 &&
    crop.y + crop.height <= 1
  )
}

function validShapeMask(mask?: EditorShapeMaskV2) {
  if (!mask) return true
  const runtimeType: string = mask.type
  return (
    (runtimeType === 'rectangle' || runtimeType === 'ellipse') &&
    [mask.x, mask.y, mask.width, mask.height].every(Number.isFinite) &&
    typeof mask.inverted === 'boolean' &&
    mask.x >= 0 &&
    mask.y >= 0 &&
    mask.width > 0 &&
    mask.height > 0 &&
    mask.x + mask.width <= 1 &&
    mask.y + mask.height <= 1
  )
}

function validPixelMask(mask?: EditorPixelMaskV2) {
  return (
    mask === undefined ||
    (/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(
      mask.resource_id,
    ) &&
      Number.isSafeInteger(mask.version) &&
      mask.version >= 0)
  )
}

function validEffects(effects: EditorEffectV2[]) {
  if (effects.length > 16) return false
  const definitions: Partial<
    Record<string, Partial<Record<string, [number, number]>>>
  > = {
    exposure: { stops: [-5, 5] },
    contrast: { amount: [-1, 1] },
    saturation: { amount: [-1, 1] },
    temperature: { kelvin_delta: [-10_000, 10_000] },
  }
  return effects.every((effect) => {
    const definition = definitions[effect.type]
    const entries = Object.entries(effect.parameters)
    return (
      definition !== undefined &&
      entries.length === Object.keys(definition).length &&
      entries.every(([key, value]) => {
        const bounds = definition[key]
        return (
          bounds !== undefined &&
          Number.isFinite(value) &&
          value >= bounds[0] &&
          value <= bounds[1]
        )
      })
    )
  })
}

function validTransform(transform: EditorTransform) {
  return (
    transform.every(
      (value) => Number.isFinite(value) && Math.abs(value) <= 1_000_000,
    ) &&
    Math.abs(transform[0] * transform[3] - transform[1] * transform[2]) >= 1e-8
  )
}

function validAncestry(
  start: EditorNodeV2,
  nodes: Map<string, EditorNodeV2>,
  key: 'parent_id' | 'mask_id',
  maximumDepth: number,
) {
  const seen = new Set([start.id])
  let depth = 1
  let nextID = start[key]
  while (nextID) {
    if (seen.has(nextID) || depth >= maximumDepth) return false
    seen.add(nextID)
    const current = nodes.get(nextID)
    if (!current) return false
    nextID = current[key]
    depth += 1
  }
  return true
}

File: backend/internal/editor/document_v2.go
package editor

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"sort"
	"strings"
	"unicode/utf8"

	"github.com/google/uuid"
)

const (
	MaxDocumentV2Bytes = 2 << 20
	MaxNodesV2         = 500
	MaxNodeDepthV2     = 32
)

var ErrUnsupportedDocumentSemantics = errors.New("unsupported editor document semantics")

// DecodedDocument is the version-neutral boundary used by persistence and
// authorization code. Rendering remains deliberately explicit: V2 documents
// must compile without loss to the currently authoritative V1 renderer.
type DecodedDocument struct {
	SchemaVersion int
	V1            *Document
	V2            *DocumentV2
}

func DecodeAny(raw []byte) (DecodedDocument, error) {
	if len(raw) == 0 || len(raw) > MaxDocumentV2Bytes {
		if len(raw) > MaxDocumentV2Bytes {
			return DecodedDocument{}, ErrDocumentTooLarge
		}
		return DecodedDocument{}, ErrInvalidDocument
	}
	var header struct {
		SchemaVersion int `json:"schema_version"`
	}
	if err := json.Unmarshal(raw, &header); err != nil {
		return DecodedDocument{}, fmt.Errorf("%w: %v", ErrInvalidDocument, err)
	}
	switch header.SchemaVersion {
	case 1:
		document, err := Decode(raw)
		if err != nil {
			return DecodedDocument{}, err
		}
		return DecodedDocument{SchemaVersion: 1, V1: &document}, nil
	case 2:
		document, err := DecodeV2(raw)
		if err != nil {
			return DecodedDocument{}, err
		}
		return DecodedDocument{SchemaVersion: 2, V2: &document}, nil
	default:
		return DecodedDocument{}, ErrInvalidDocument
	}
}

func DecodeRenderable(raw []byte) (Document, error) {
	document, err := DecodeAny(raw)
	if err != nil {
		return Document{}, err
	}
	return document.RenderableV1()
}

func (d DecodedDocument) RenderableV1() (Document, error) {
	switch d.SchemaVersion {
	case 1:
		if d.V1 == nil {
			return Document{}, ErrInvalidDocument
		}
		return *d.V1, nil
	case 2:
		if d.V2 == nil {
			return Document{}, ErrInvalidDocument
		}
		return d.V2.ToV1()
	default:
		return Document{}, ErrInvalidDocument
	}
}

func (d DecodedDocument) AssetIDs() []uuid.UUID {
	switch d.SchemaVersion {
	case 1:
		if d.V1 != nil {
			return d.V1.AssetIDs()
		}
	case 2:
		if d.V2 != nil {
			return d.V2.AssetIDs()
		}
	}
	return nil
}

func (d DecodedDocument) PixelMaskReferences() []PixelMaskReferenceV2 {
	if d.SchemaVersion != 2 || d.V2 == nil {
		return nil
	}
	references := make([]PixelMaskReferenceV2, 0)
	for _, node := range d.V2.Nodes {
		if node.PixelMask == nil {
			continue
		}
		references = append(references, PixelMaskReferenceV2{
			NodeID: node.ID, AssetID: *node.AssetID,
			ResourceID: node.PixelMask.ResourceID, Version: node.PixelMask.Version,
		})
	}
	return references
}

type DocumentV2 struct {
	SchemaVersion            int      `json:"schema_version"`
	RendererSemanticsVersion int      `json:"renderer_semantics_version"`
	Canvas                   Canvas   `json:"canvas"`
	Nodes                    []NodeV2 `json:"nodes"`
}

type NodeV2 struct {
	ID        string       `json:"id"`
	Type      string       `json:"type"`
	Name      string       `json:"name,omitempty"`
	ParentID  *string      `json:"parent_id"`
	OrderKey  string       `json:"order_key"`
	Transform [6]float64   `json:"transform"`
	Opacity   float64      `json:"opacity"`
	BlendMode string       `json:"blend_mode"`
	Visible   bool         `json:"visible"`
	Locked    bool         `json:"locked"`
	MaskID    *string      `json:"mask_id,omitempty"`
	AssetID   *uuid.UUID   `json:"asset_id,omitempty"`
	Crop      *Crop        `json:"crop,omitempty"`
	Effects   []EffectV2   `json:"effects"`
	TargetID  *string      `json:"target_id,omitempty"`
	ShapeMask *ShapeMaskV2 `json:"shape_mask,omitempty"`
	PixelMask *PixelMaskV2 `json:"pixel_mask,omitempty"`
}

type PixelMaskV2 struct {
	ResourceID uuid.UUID `json:"resource_id"`
	Version    int64     `json:"version"`
}

type PixelMaskReferenceV2 struct {
	NodeID     string
	AssetID    uuid.UUID
	ResourceID uuid.UUID
	Version    int64
}

type ShapeMaskV2 struct {
	Type     string  `json:"type"`
	X        float64 `json:"x"`
	Y        float64 `json:"y"`
	Width    float64 `json:"width"`
	Height   float64 `json:"height"`
	Inverted bool    `json:"inverted"`
}

type EffectV2 struct {
	Type       string             `json:"type"`
	Version    int                `json:"version"`
	Enabled    bool               `json:"enabled"`
	Parameters map[string]float64 `json:"parameters"`
}

func DecodeV2(raw []byte) (DocumentV2, error) {
	if len(raw) == 0 || len(raw) > MaxDocumentV2Bytes {
		if len(raw) > MaxDocumentV2Bytes {
			return DocumentV2{}, ErrDocumentTooLarge
		}
		return DocumentV2{}, ErrInvalidDocument
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	var document DocumentV2
	if err := decoder.Decode(&document); err != nil {
		return DocumentV2{}, fmt.Errorf("%w: %v", ErrInvalidDocument, err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return DocumentV2{}, ErrInvalidDocument
	}
	if err := document.Validate(); err != nil {
		return DocumentV2{}, err
	}
	return document, nil
}

func MigrateV1ToV2(document Document) (DocumentV2, error) {
	if err := document.Validate(); err != nil {
		return DocumentV2{}, err
	}
	objects := append([]Object(nil), document.Objects...)
	sort.SliceStable(objects, func(i, j int) bool { return objects[i].ZIndex < objects[j].ZIndex })
	nodes := make([]NodeV2, len(objects))
	for index, object := range objects {
		assetID := object.AssetID
		nodes[index] = NodeV2{
			ID: object.ID, Type: "raster", Name: object.Name,
			OrderKey: fmt.Sprintf("%08d", index), Transform: object.Transform,
			Opacity: object.Opacity, BlendMode: "normal", Visible: object.Visible, Locked: object.Locked,
			AssetID: &assetID, Crop: object.Crop, Effects: []EffectV2{},
		}
	}
	return DocumentV2{
		SchemaVersion: 2, RendererSemanticsVersion: 1,
		Canvas: document.Canvas, Nodes: nodes,
	}, nil
}

func (d DocumentV2) ToV1() (Document, error) {
	if err := d.Validate(); err != nil {
		return Document{}, err
	}
	nodes := append([]NodeV2(nil), d.Nodes...)
	for _, node := range nodes {
		if node.Type != "raster" || node.ParentID != nil || node.MaskID != nil || node.ShapeMask != nil || node.PixelMask != nil || node.BlendMode != "normal" || len(node.Effects) != 0 || node.AssetID == nil {
			return Document{}, ErrUnsupportedDocumentSemantics
		}
	}
	sort.SliceStable(nodes, func(i, j int) bool {
		if nodes[i].OrderKey == nodes[j].OrderKey {
			return nodes[i].ID < nodes[j].ID
		}
		return nodes[i].OrderKey < nodes[j].OrderKey
	})
	objects := make([]Object, len(nodes))
	for index, node := range nodes {
		objects[index] = Object{
			ID: node.ID, Name: node.Name, AssetID: *node.AssetID, Transform: node.Transform,
			Opacity: node.Opacity, Visible: node.Visible, Locked: node.Locked, ZIndex: index, Crop: node.Crop,
		}
	}
	document := Document{SchemaVersion: 1, Canvas: d.Canvas, Objects: objects}
	if err := document.Validate(); err != nil {
		return Document{}, err
	}
	return document, nil
}

func (d DocumentV2) Validate() error {
	if d.SchemaVersion != 2 || d.RendererSemanticsVersion != 1 || !validCanvas(d.Canvas) || len(d.Nodes) < 1 || len(d.Nodes) > MaxNodesV2 {
		return ErrInvalidDocument
	}
	nodes := make(map[string]NodeV2, len(d.Nodes))
	orderKeys := make(map[string]struct{}, len(d.Nodes))
	for _, node := range d.Nodes {
		if !validNodeID(node.ID) || !validNodeName(node.Name) || !validOrderKey(node.OrderKey) || !validTransform(node.Transform) || !validOpacity(node.Opacity) || !validBlendMode(node.BlendMode) {
			return ErrInvalidDocument
		}
		if _, exists := nodes[node.ID]; exists {
			return ErrInvalidDocument
		}
		parent := "\x00"
		if node.ParentID != nil {
			if *node.ParentID == node.ID {
				return ErrInvalidDocument
			}
			parent = "\x01" + *node.ParentID
		}
		orderIdentity := parent + "\x00" + node.OrderKey
		if _, exists := orderKeys[orderIdentity]; exists {
			return ErrInvalidDocument
		}
		orderKeys[orderIdentity] = struct{}{}
		if node.Type != "raster" && node.Type != "group" && node.Type != "adjustment" {
			return ErrInvalidDocument
		}
		if node.Type == "raster" {
			if node.AssetID == nil || *node.AssetID == uuid.Nil || node.TargetID != nil || !validCrop(node.Crop) || !validEffects(node.Effects) || !validShapeMaskV2(node.ShapeMask) || !validPixelMaskV2(node.PixelMask) || (node.ShapeMask != nil && (node.Crop != nil || node.MaskID != nil || node.PixelMask != nil)) || (node.PixelMask != nil && node.MaskID != nil) {
				return ErrInvalidDocument
			}
		} else if node.Type == "group" {
			if node.AssetID != nil || node.Crop != nil || node.TargetID != nil || node.ShapeMask != nil || node.PixelMask != nil || len(node.Effects) != 0 {
				return ErrInvalidDocument
			}
		} else if node.AssetID != nil || node.Crop != nil || node.MaskID != nil || node.ShapeMask != nil || node.PixelMask != nil || node.TargetID == nil || *node.TargetID == node.ID || node.Transform != [6]float64{1, 0, 0, 1, 0, 0} || node.BlendMode != "normal" || !validEffects(node.Effects) {
			return ErrInvalidDocument
		}
		nodes[node.ID] = node
	}
	for _, node := range d.Nodes {
		if node.ParentID != nil {
			parent, exists := nodes[*node.ParentID]
			if !exists || parent.Type != "group" {
				return ErrInvalidDocument
			}
		}
		if node.MaskID != nil {
			mask, exists := nodes[*node.MaskID]
			if !exists || mask.Type != "raster" || mask.ID == node.ID {
				return ErrInvalidDocument
			}
		}
		if node.Type == "adjustment" {
			target, exists := nodes[*node.TargetID]
			if !exists || target.Type != "raster" || target.ParentID == nil != (node.ParentID == nil) || (target.ParentID != nil && *target.ParentID != *node.ParentID) {
				return ErrInvalidDocument
			}
			for _, candidate := range d.Nodes {
				if candidate.MaskID != nil && *candidate.MaskID == target.ID {
					return ErrInvalidDocument
				}
			}
		}
		if node.ShapeMask != nil {
			for _, candidate := range d.Nodes {
				if candidate.MaskID != nil && *candidate.MaskID == node.ID {
					return ErrInvalidDocument
				}
			}
		}
		if node.PixelMask != nil {
			for _, candidate := range d.Nodes {
				if candidate.MaskID != nil && *candidate.MaskID == node.ID {
					return ErrInvalidDocument
				}
			}
		}
		if depth, ok := nodeDepth(node, nodes); !ok || depth > MaxNodeDepthV2 {
			return ErrInvalidDocument
		}
		if !validMaskChain(node, nodes) {
			return ErrInvalidDocument
		}
	}
	return nil
}

func (d DocumentV2) AssetIDs() []uuid.UUID {
	ids := make([]uuid.UUID, 0, len(d.Nodes))
	seen := make(map[uuid.UUID]struct{}, len(d.Nodes))
	for _, node := range d.Nodes {
		if node.AssetID == nil {
			continue
		}
		if _, exists := seen[*node.AssetID]; exists {
			continue
		}
		seen[*node.AssetID] = struct{}{}
		ids = append(ids, *node.AssetID)
	}
	return ids
}

func validCanvas(canvas Canvas) bool {
	return canvas.Width >= 1 && canvas.Height >= 1 && canvas.Width <= 8192 && canvas.Height <= 8192 && int64(canvas.Width)*int64(canvas.Height) <= MaxCanvasPixels
}

func validNodeID(value string) bool {
	return len(value) >= 1 && len(value) <= 64 && strings.TrimSpace(value) == value && utf8.ValidString(value)
}

func validNodeName(value string) bool {
	return len(value) <= 256 && utf8.ValidString(value) && utf8.RuneCountInString(value) <= 64
}

func validOrderKey(value string) bool {
	if len(value) < 1 || len(value) > 64 {
		return false
	}
	for _, char := range value {
		if !((char >= '0' && char <= '9') || (char >= 'A' && char <= 'Z') || (char >= 'a' && char <= 'z')) {
			return false
		}
	}
	return true
}

func validTransform(transform [6]float64) bool {
	for _, value := range transform {
		if math.IsNaN(value) || math.IsInf(value, 0) || math.Abs(value) > 1_000_000 {
			return false
		}
	}
	return math.Abs(transform[0]*transform[3]-transform[1]*transform[2]) >= 1e-8
}

func validOpacity(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0) && value >= 0 && value <= 1
}

func validBlendMode(value string) bool {
	switch value {
	case "normal", "multiply", "screen", "overlay", "darken", "lighten":
		return true
	default:
		return false
	}
}

func validCrop(crop *Crop) bool {
	if crop == nil {
		return true
	}
	sum := crop.X + crop.Y + crop.Width + crop.Height
	return crop.X >= 0 && crop.Y >= 0 && crop.Width > 0 && crop.Height > 0 && crop.X+crop.Width <= 1 && crop.Y+crop.Height <= 1 && !math.IsNaN(sum) && !math.IsInf(sum, 0)
}

func validShapeMaskV2(mask *ShapeMaskV2) bool {
	if mask == nil {
		return true
	}
	sum := mask.X + mask.Y + mask.Width + mask.Height
	return (mask.Type == "rectangle" || mask.Type == "ellipse") && mask.X >= 0 && mask.Y >= 0 && mask.Width > 0 && mask.Height > 0 && mask.X+mask.Width <= 1 && mask.Y+mask.Height <= 1 && !math.IsNaN(sum) && !math.IsInf(sum, 0)
}

func validPixelMaskV2(mask *PixelMaskV2) bool {
	return mask == nil || (mask.ResourceID != uuid.Nil && mask.Version >= 0)
}

func validEffects(effects []EffectV2) bool {
	if len(effects) > 16 {
		return false
	}
	for _, effect := range effects {
		if effect.Version != 1 || len(effect.Parameters) == 0 {
			return false
		}
		var allowed map[string][2]float64
		switch effect.Type {
		case "exposure":
			allowed = map[string][2]float64{"stops": {-5, 5}}
		case "contrast", "saturation":
			allowed = map[string][2]float64{"amount": {-1, 1}}
		case "temperature":
			allowed = map[string][2]float64{"kelvin_delta": {-10000, 10000}}
		default:
			return false
		}
		if len(effect.Parameters) != len(allowed) {
			return false
		}
		for key, value := range effect.Parameters {
			bounds, exists := allowed[key]
			if !exists || math.IsNaN(value) || math.IsInf(value, 0) || value < bounds[0] || value > bounds[1] {
				return false
			}
		}
	}
	return true
}

func nodeDepth(node NodeV2, nodes map[string]NodeV2) (int, bool) {
	seen := map[string]struct{}{node.ID: {}}
	depth := 1
	for node.ParentID != nil {
		if _, exists := seen[*node.ParentID]; exists {
			return 0, false
		}
		seen[*node.ParentID] = struct{}{}
		parent, exists := nodes[*node.ParentID]
		if !exists {
			return 0, false
		}
		node = parent
		depth++
	}
	return depth, true
}

func validMaskChain(node NodeV2, nodes map[string]NodeV2) bool {
	seen := map[string]struct{}{node.ID: {}}
	for node.MaskID != nil {
		if _, exists := seen[*node.MaskID]; exists {
			return false
		}
		seen[*node.MaskID] = struct{}{}
		node = nodes[*node.MaskID]
	}
	return true
}
